spec-0012/index.md

```diff
@@ -4,122 +4,180 @@ number: 12
-e.g. `a + b + c` is a single group, `a + b * c` is composed of two groups `a`
-and `b * c`. A group is also a collection delimited with parenthesis.
-`(a + b * c)` is a group. And the whole expression by itself is a
-group. 
-
-- There a space before and after `-` and `+`. Except if
-  the operator is used to define the sign of the number;
-
-  ```
-  a + b
-  -a
-  ```
-
-- Within a group, if operators with different priorities are used, add whitespace around the operators with the lowest priority(ies).
-
-  ```
-  a + b*c
-  ```
-
-- There is no space before and after `**`.
-
-  ```
-  a**b
-  ```
-
-- There is no space before and after operators `*` and `/`. Only exception is if the expression consist of a single operator linking two groups with more than one
- element.
+## Terminology
 
-  ```
-  a*b
-  (a*b) * (c*d)
-  ```
-
-- Operators within a group are ordered from the lowest to the highest priority.
-  If this is technically an issue (e.g. restriction on the AST), add
-  parenthesis or spaces.
-
-  ```
-  a/d*b**c
-  a*(b**c)/d
-  a*b**c / d
-  a * b**c / d
-  ```
-
-- When splitting an equation, new lines should start with the operator linking
-  the previous and next logical block. Single digit on a line are forbidden.
-
-  ```
-  (
-      a/b
-      + c*d
-  )
-  ```
-
-### Examples
-
-```python
-# good
-i = i + 1
-submitted += 1
-x = x*2 - 1
-hypot2 = x*x + y*y
-c = (a + b) * (a - b)
-dfdx = sign*(-2*x + 2*y + 2)
-result = 2*x**2 + 3*x**(2/3)
-y = 4*x**2 + 2*x + 1
-c_i1j = (
-    1./n**2.
-    *np.prod(
-        0.5*(2. + abs(z_ij[i1, :]) + abs(z_ij) - abs(z_ij[i1, :] - z_ij)), axis=1
-    )
-)
-```
-
-```python
-# bad
-i = i + 1
-submitted += 1
-x = x * 2 - 1
-hypot2 = x * x + y * y
-c = (a + b) * (a - b)
-dfdx = sign * (-2 * x + 2 * y + 2)
-result = 2 * x ** 2 + 3 * x ** (2 / 3)
-y = 4 * x ** 2 + 2 * x + 1
-c_i1j = (
-    1.0
-    / n ** 2.0
-    * np.prod(
-        0.5 * (2.0 + abs(z_ij[i1, :]) + abs(z_ij) - abs(z_ij[i1, :] - z_ij)), axis=1
-    )
-)
+An "explicit" expression is a code expression enclosed within parentheses or
+otherwise syntactically separated from other expressions (i.e. by code other
+than operators, whitespace, literals, or variables). For example, in the list
+comprehension:
+```python3
+for j in range(1, i + 1)
 ```
+The output expression `j` is one explicit expression and the input sequence
+`range(1, i + 1)` is another. 
+
+A "subexpression" is subset of an expression that is either explicit or could
+be made explicit (i.e. with parentheses) without affecting the order of
+operations. In the example above, `j` and `range(1, i + 1)` can also be
+referred to as explicit subexpressions of  the whole expression, and `1` and
+`i + 1` are explicit subexpressions of the expression `range(1, i + 1)`. `i` and
+`1` are "implicit" subexpressions of `i + 1`: they could be written as explicit
+subexpressions `(i)` and `(1)` without affecting the order  of operations, but they
+are not explicit as written.
```

Comment:
I think this concept might still need refinement to make the rules unambiguous, but let's see if others find ambiguities before adding unnecessarily.